torchmetrics/audio/sdr.py

```diff
@@ -0,0 +1,138 @@
+
+from torchmetrics.functional.audio.sdr import sdr
+from torchmetrics.metric import Metric
+
+
+class SDR(Metric):
+    r"""Signal to Distortion Ratio (SDR) [1,2,3]
+
+    .. note:: 1) using this metrics requires you to have ``fast-bss-eval`` install. Either install as ``pip install
+        torchmetrics[audio]`` or ``pip install fast-bss-eval``
+        2) preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
+        3) when pytorch<1.8.0, numpy will be used to calculate this metric, which causes ``sdr`` non-differentiable
+
+    Forward accepts
+
+    - ``preds``: shape ``[..., time]``
+    - ``target``: shape ``[..., time]``
+
+    Args:
+        use_cg_iter:
+            If provided, an iterative method is used to solve for the distortion \
+            filter coefficients instead of direct Gaussian elimination. \
+            This can speed up the computation of the metrics in case the filters \
+            are long. Using a value of 10 here has been shown to provide \
+            good accuracy in most cases and is sufficient when using this \
+            loss to train neural separation networks.
+        filter_length:
+            The length of the distortion filter allowed
+        zero_mean:
+            When set to True, the mean of all signals is subtracted prior to computation of the metrics
+        load_diag:
+            If provided, this small value is added to the diagonal coefficients of \
+            the system metrics when solving for the filter coefficients. \
+            This can help stabilize the metric in the case where some of the reference \
+            signals may sometimes be zero
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step.
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When `None`, DDP
+            will be used to perform the allgather.
+
+    Raises:
+        ValueError:
+            If ``fast-bss-eval`` package is not installed
+
+    Example:
+        >>> from torchmetrics.audio import SDR
+        >>> import torch
+        >>> g = torch.manual_seed(1)
+        >>> preds = torch.randn(8000)
+        >>> target = torch.randn(8000)
+        >>> sdr = SDR()
+        >>> sdr(preds, target)
+        tensor(-12.0589)
+        >>> # use with pit
+        >>> from torchmetrics.audio import PIT
+        >>> from torchmetrics.functional.audio import sdr
+        >>> # [batch, spk, time]
+        >>> preds = torch.randn(4, 2, 8000)
+        >>> target = torch.randn(4, 2, 8000)
+        >>> pit = PIT(sdr, 'max')
```

Comment:
why PIT not SDR?

Reply:
It's a use case for SDR with pit, to calculate sdr in a permutation invariant way